### PR TITLE
feat: implement sever-side stream changes to gapic-common

### PR DESCRIPTION
**Pull request by atulep**
Implements changes for server-side streaming to gapic-common.

### CODE REVIEW

File: gapic-common/lib/gapic/rest/client_stub.rb [hunk 1]
```diff
@@ -127,8 +141,11 @@ def make_put_request uri:, body: nil, params: {}, options: {}
       # @param params [Hash] query string parameters for the request
       # @param options [::Gapic::CallOptions,Hash] gapic options to be applied
       #     to the REST call. Currently only timeout and headers are supported.
+      # @param is_server_streaming [Bool] flag if method is streaming
```
Comment (viacheslav-rostovtsev) [inline gapic-common/lib/gapic/rest/client_stub.rb:144]: "is server setreaming" in the param description

File: gapic-common/lib/gapic/rest/threaded_enumerator.rb [hunk 2]
```diff
@@ -0,0 +1,68 @@
...
+    # ThreadedEnumerator provides the enumerations over the individual chunks of data received from the server.
+    #
+    # @example normal iteration over resources.
+    #   chunk = threaded_enumerator.next
+    #
+    # @attribute [r] in_q
+    #   @return [Queue] Input queue.
+    # @attribute [r] out_q
+    #   @return [Queue] Output queue.
+    class ThreadedEnumerator
```
Comment (viacheslav-rostovtsev) [inline gapic-common/lib/gapic/rest/threaded_enumerator.rb:30]: let's make it private

File: gapic-common/lib/gapic/rest/client_stub.rb [hunk 3]
```diff
@@ -118,6 +118,20 @@ def make_put_request uri:, body: nil, params: {}, options: {}
         make_http_request :put, uri: uri, body: body, params: params, options: options
       end
 
+      ##
+      # @private
+      # Yields to the block if method is streaming.
+      # @param req [Faraday::Request] request object
+      # @param is_server_streaming [Bool] flag if method is streaming
+      # @yield None
+      # @yieldparam chunk[string] The chunk of data received during server streaming.
```
Comment (viacheslav-rostovtsev) [inline gapic-common/lib/gapic/rest/client_stub.rb:127]: merge back

File: gapic-common/lib/gapic/rest/server_stream.rb [hunk 4]
```diff
@@ -0,0 +1,92 @@
...
+              return
+            end
+          end
+          yield @message_klass.decode_json @ready_objs.shift, ignore_unknown_fields: true
+        end
+      end
+
+      private
+
+      def _next_json! chunk
```
Comment (viacheslav-rostovtsev) [inline gapic-common/lib/gapic/rest/server_stream.rb:65]: add types to comment

File: gapic-common/lib/gapic/rest/client_stub.rb (hunk 3 above)
Comment (viacheslav-rostovtsev) [inline gapic-common/lib/gapic/rest/client_stub.rb:127]: String

File: gapic-common/lib/gapic/rest/threaded_enumerator.rb (hunk 2 above)
Comment (atulep) [reply gapic-common/lib/gapic/rest/threaded_enumerator.rb:30]: I guessed you're meant making attributes private.

File: gapic-common/lib/gapic/rest/client_stub.rb (hunk 3 above)
Comment (atulep) [reply gapic-common/lib/gapic/rest/client_stub.rb]: Done

File: gapic-common/lib/gapic/rest/server_stream.rb (hunk 4 above)
Comment (atulep) [reply gapic-common/lib/gapic/rest/server_stream.rb:77]: Done

File: gapic-common/lib/gapic/rest/client_stub.rb (hunk 3 above)
Comment (atulep) [reply gapic-common/lib/gapic/rest/client_stub.rb]: Done

File: gapic-common/lib/gapic/rest/client_stub.rb (hunk 1 above)
Comment (atulep) [reply gapic-common/lib/gapic/rest/client_stub.rb:130]: Done

File: gapic-common/lib/gapic/rest/threaded_enumerator.rb (hunk 2 above)
Comment (atulep) [reply gapic-common/lib/gapic/rest/threaded_enumerator.rb:31]: Done

File: gapic-common/lib/gapic/rest/server_stream.rb [hunk 5]
```diff
@@ -0,0 +1,104 @@
...
+        chunk.chars.each do |char|
+          @_obj += char
+          # Invariant: @_obj is always either a part of a single JSON object or the entire JSON object.
+          # Hence, it's safe to strip whitespace, commans and array brackets. These characters
+          # are only added before @_obj is a complete JSON object and essentially can be flushed.
+          @_obj = @_obj.lstrip # strip whitespace.
+          # Eat array delimiter characters.
+          if @_obj[0] == "[" || @_obj[0] == "," || @_obj[0] == "]"
+            @_obj = @_obj[1..]
+          end
```
Comment (dazuma) [inline gapic-common/lib/gapic/rest/server_stream.rb:87]: I think we can optimize this a lot better. Currently we're doing a lot of calls, including creating multiple new objects and copying a lot of string data, on every character added. This makes for a very slow inner loop. Can we instead:

* Replace the logic in lines 83-87 with a single call like `@_obj.sub!(/^([\s\[\],)+/, "")`
* Lift that call out of the loop. It should only need to happen once, to flush out those characters before we start reading new ones.
* Then split the each loop into two phases: the first that reads characters in the skip list (space, brackets, and commas) and throws them away, and then the second that starts with the first real character and does what you're doing below.

File: gapic-common/lib/gapic/rest/threaded_enumerator.rb [hunk 6]
```diff
@@ -0,0 +1,73 @@
...
+      attr_reader :in_q
+      attr_reader :out_q
+
+      # Spawns a new thread and does appropriate clean-up
+      # in case thread fails. Propagates exception back
+      # to main thread.
+      #
+      # @yield None
```
Comment (dazuma) [inline gapic-common/lib/gapic/rest/threaded_enumerator.rb:39]: Delete this line. Not needed.

File: gapic-common/lib/gapic/rest/threaded_enumerator.rb [hunk 7]
```diff
@@ -0,0 +1,73 @@
...
+      # @yieldparam in_q[Queue] input queue
+      # @yieldparam out_q[Queue] output queue
+      def initialize &block
+        @in_q = Queue.new
+        @out_q = Queue.new
+        @block = block
```
Comment (dazuma) [inline gapic-common/lib/gapic/rest/threaded_enumerator.rb:45]: Delete this line. `@block` is never used outside the constructor.

File: gapic-common/lib/gapic/rest/threaded_enumerator.rb [hunk 8]
```diff
@@ -0,0 +1,73 @@
...
+
+        Thread.new do
+          @block.call @in_q, @out_q
```
Comment (dazuma) [inline gapic-common/lib/gapic/rest/threaded_enumerator.rb:48]: Just `yield @in_q, @out_q`

**Review by dazuma — CHANGES_REQUESTED** [review]

File: gapic-common/lib/gapic/rest/threaded_enumerator.rb (hunk 8 above)
Comment (atulep) [reply gapic-common/lib/gapic/rest/threaded_enumerator.rb]: Done

File: gapic-common/lib/gapic/rest/threaded_enumerator.rb (hunk 7 above)
Comment (atulep) [reply gapic-common/lib/gapic/rest/threaded_enumerator.rb]: Done.

File: gapic-common/lib/gapic/rest/threaded_enumerator.rb (hunk 6 above)
Comment (atulep) [reply gapic-common/lib/gapic/rest/threaded_enumerator.rb]: Done

File: gapic-common/lib/gapic/rest/server_stream.rb [hunk 9]
```diff
@@ -0,0 +1,104 @@
...
+              next unless chunk
+              _next_json! chunk
+            rescue StopIteration
+              dangling_content = @_obj.strip
+              error_expl = "Dangling conent left after iterating through the stream. " \
+                           "This means that not all content was received or parsed correctly. " \
+                           "It is likely a result of server or network error."
+              error_text = "#{error_expl}\n Content left unparsed: #{dangling_content}"
+
+              raise Gapic::Common::Error, error_text unless dangling_content.empty?
```
[the remark this answers is not in the document]
Comment (atulep) [reply gapic-common/lib/gapic/rest/server_stream.rb]: Done.

File: gapic-common/lib/gapic/rest/client_stub.rb [hunk 10]
```diff
@@ -127,8 +127,11 @@ def make_put_request uri:, body: nil, params: {}, options: {}
       # @param params [Hash] query string parameters for the request
       # @param options [::Gapic::CallOptions,Hash] gapic options to be applied
       #     to the REST call. Currently only timeout and headers are supported.
+      # @param is_server_streaming [Bool] flag if method is streaming
+      # @yield None
+      # @yieldparam chunk[String] The chunk of data received during server streaming.
```
[the remark this answers is not in the document]
Comment (atulep) [reply gapic-common/lib/gapic/rest/client_stub.rb]: Done

File: gapic-common/lib/gapic/rest/client_stub.rb [hunk 11]
```diff
@@ -127,8 +127,11 @@ def make_put_request uri:, body: nil, params: {}, options: {}
       # @param params [Hash] query string parameters for the request
       # @param options [::Gapic::CallOptions,Hash] gapic options to be applied
       #     to the REST call. Currently only timeout and headers are supported.
+      # @param is_server_streaming [Bool] flag if method is streaming
+      # @yield None
```
[the remark this answers is not in the document]
Comment (atulep) [reply gapic-common/lib/gapic/rest/client_stub.rb]: Done

File: gapic-common/lib/gapic/rest/client_stub.rb [hunk 12]
```diff
@@ -127,8 +127,11 @@ def make_put_request uri:, body: nil, params: {}, options: {}
       # @param params [Hash] query string parameters for the request
       # @param options [::Gapic::CallOptions,Hash] gapic options to be applied
       #     to the REST call. Currently only timeout and headers are supported.
+      # @param is_server_streaming [Bool] flag if method is streaming
```
[the remark this answers is not in the document]
Comment (atulep) [reply gapic-common/lib/gapic/rest/client_stub.rb]: Done

File: gapic-common/lib/gapic/rest/client_stub.rb [hunk 13]
```diff
@@ -138,6 +141,11 @@ def make_http_request verb, uri:, body:, params:, options:
           req.body = body unless body.nil?
           req.headers = req.headers.merge options.metadata
           req.options.timeout = options.timeout if options.timeout&.positive?
+          if is_server_streaming
+            req.options.on_data = proc do |chunk, _overall_received_bytes|
+              block.call chunk if block_given?
```
[the remark this answers is not in the document]
Comment (atulep) [reply gapic-common/lib/gapic/rest/client_stub.rb]: Ah, that's clever. Changed.

[the remark this answers is not in the document]
Comment (atulep) [reply gapic-common/lib/gapic/rest/client_stub.rb]: I removed block but kept server streaming. I think there might be other ways block get used outside of server-streaming (i.e. client streaming) and don't want to bind its use at this time.

[the remark this answers is not in the document]
Comment (atulep) [reply gapic-common/lib/gapic/rest/client_stub.rb]: Can also raise if is_server_streaming is true and block not provided as alternative.

File: gapic-common/lib/gapic/rest/server_stream.rb (hunk 5 above)
Comment (atulep) [reply gapic-common/lib/gapic/rest/server_stream.rb:88]: Thanks for the suggestion. I don't understand the last part. Do you mean split each loop into two each loops ?

File: gapic-common/lib/gapic/rest/server_stream.rb [hunk 14]
```diff
@@ -0,0 +1,104 @@
...
+    #   server_stream.each { |response| puts response }
+    #
+    class ServerStream
+      include Enumerable
+
+      # @param message_klass [Class]
+      # @param json_enumerator [Enumerator<String>]
+      def initialize message_klass, json_enumerator
+        @json_enumerator = json_enumerator
+        @_obj = ""
```
[the remark this answers is not in the document]
Comment (atulep) [reply gapic-common/lib/gapic/rest/server_stream.rb]: Done

File: gapic-common/lib/gapic/rest/server_stream.rb [hunk 15]
```diff
@@ -0,0 +1,104 @@
...
+
+        loop do
+          while @ready_objs.length.zero?
+            begin
+              chunk = @json_enumerator.next
+              next unless chunk
+              _next_json! chunk
+            rescue StopIteration
+              dangling_content = @_obj.strip
+              error_expl = "Dangling conent left after iterating through the stream. " \
```
[the remark this answers is not in the document]
Comment (atulep) [reply gapic-common/lib/gapic/rest/server_stream.rb]: Done

File: gapic-common/lib/gapic/rest/server_stream.rb [hunk 16]
```diff
@@ -0,0 +1,104 @@
...
+      end
+
+      private
+
+      ##
+      # Builds the next JSON object of the server stream from chunk.
+      #
+      # @param chunk [String] Contains (partial) JSON object
+      #
+      def _next_json! chunk
```
[the remark this answers is not in the document]
Comment (atulep) [reply gapic-common/lib/gapic/rest/server_stream.rb]: Done.

File: gapic-common/lib/gapic/rest/server_stream.rb [hunk 17]
```diff
@@ -0,0 +1,104 @@
...
+    #
+    # ServerStream provides the enumerations over the individual response messages within the stream.
+    #
+    # @example normal iteration over resources.
+    #   server_stream.each { |response| puts response }
+    #
+    class ServerStream
+      include Enumerable
+
+      # @param message_klass [Class]
```
[the remark this answers is not in the document]
Comment (atulep) [reply gapic-common/lib/gapic/rest/server_stream.rb:33]: Done.

File: gapic-common/lib/gapic/rest/client_stub.rb (hunk 13 above)
[the remark this answers is not in the document]
Comment (atulep) [reply gapic-common/lib/gapic/rest/client_stub.rb]: Using yield.

File: gapic-common/lib/gapic/rest/server_stream.rb (hunk 5 above)
Comment (dazuma) [reply gapic-common/lib/gapic/rest/server_stream.rb]: Something else you might consider is: `chunk.scan(/\{|\}|[^{}]+/) do |chrs| ...`. That would reduce the number of iterations significantly, by combining runs of characters that aren't braces.

Comment (atulep) [reply gapic-common/lib/gapic/rest/server_stream.rb:87]: Thanks so much for thorough explanation. I added an optimization you suggested to skip unneccesary string concatenations. I used `char.strip` to check for whitespaces - do you think it's slow?

Comment (dazuma) [reply gapic-common/lib/gapic/rest/server_stream.rb:87]: Following up: I suggest `next if @obj.empty? && char != "{"` to replace lines 85-87.

Comment (atulep) [reply gapic-common/lib/gapic/rest/server_stream.rb]: Incredible. Thanks.

**Review by dazuma — APPROVED** [review]
LGTM

**Review by viacheslav-rostovtsev — APPROVED** [review]
LGTM

**Pull request merged**